lib/agent.js

```diff
@@ -0,0 +1,227 @@
+'use strict'
+
+const {
+  ClientClosedError,
+  InvalidArgumentError,
+  ClientDestroyedError
+} = require('./core/errors')
+const Dispatcher = require('./dispatcher')
+const Pool = require('./pool')
+const Client = require('./client')
+const util = require('./core/util')
+const assert = require('assert')
+const RedirectHandler = require('./handler/redirect')
+
+const kDestroyed = Symbol('destroyed')
+const kClosed = Symbol('closed')
+const kOnConnect = Symbol('onConnect')
+const kOnDisconnect = Symbol('onDisconnect')
+const kClients = Symbol('clients')
+const kMaxRedirections = Symbol('maxRedirections')
+const kOnDrain = Symbol('onDrain')
+const kFactory = Symbol('factory')
+const kFinalizer = Symbol('finalizer')
+const kOptions = Symbol('options')
+
+function defaultFactory (origin, opts) {
+  return opts && opts.connections === 1
+    ? new Client(origin, opts)
+    : new Pool(origin, opts)
+}
+
+const WeakRef = global.WeakRef || class CompatWeakRef {
+  constructor (value) {
+    this.value = value
+  }
+
+  deref () {
+    return this.value.connected === 0 && this.value.size === 0
```

Comment:
I think you are missing https://github.com/nodejs/undici/pull/620/files#diff-d8bb2c7932a22406705bd0c42d2cb3d720574c245ba2cc81993a7d8d563b0e3eR50

Reply:
I don't see a problem here?

Reply:
I was thinking that the finalizer is called when `connected` and `size` are `0`. In the finalized then the `WeakRef` in unrefed, and this operation will return `undefined` because it performs the same check, resulting in not removing the `dispatcher` from  the map.

Reply:
I am referring to this `if` branch:

https://github.com/nodejs/undici/pull/620/files#diff-d8bb2c7932a22406705bd0c42d2cb3d720574c245ba2cc81993a7d8d563b0e3eR76

Reply:
>will return undefined because it performs the same check, resulting in not removing the dispatcher from the map.

But it does remove it when and only when it is undefined.

Reply:
🤦 Sorry. I don't know what I was thinking. You're right!

Reply:
```js
ref.deref() === undefined
```

Reply:
Thanks